src/shared/loop-util.c

```diff
@@ -14,20 +14,65 @@
 #include <unistd.h>
 
 #include "alloc-util.h"
+#include "blockdev-util.h"
 #include "errno-util.h"
 #include "fd-util.h"
 #include "fileio.h"
 #include "loop-util.h"
+#include "missing_loop.h"
 #include "parse-util.h"
 #include "stat-util.h"
 #include "stdio-util.h"
 #include "string-util.h"
 
 static void cleanup_clear_loop_close(int *fd) {
-        if (*fd >= 0) {
-                (void) ioctl(*fd, LOOP_CLR_FD);
-                (void) safe_close(*fd);
+        if (*fd < 0)
+                return;
+
+        (void) ioctl(*fd, LOOP_CLR_FD);
+        (void) safe_close(*fd);
+}
+
+static int loop_configure(int fd, const struct loop_config *c) {
+        int r;
+
+        assert(fd >= 0);
+        assert(c);
+
+        if (ioctl(fd, LOOP_CONFIGURE, c) < 0) {
```

Comment:
This returns EINVAL when fd points to a single filesystem

Reply:
Can confirm on 5.6.2-300.fc32:

```
[   14.109806] testsuite-50.sh[620]: openat(AT_FDCWD, "/dev/loop-control", O_RDWR|O_NOCTTY|O_NONBLOCK|O_CLOEXEC) = 4
[   14.185683] kernel: loop: module loaded
[   14.158634] testsuite-50.sh[620]: ioctl(4, LOOP_CTL_GET_FREE)             = 0
[   14.158712] testsuite-50.sh[620]: openat(AT_FDCWD, "/dev/loop0", O_RDONLY|O_NOCTTY|O_NONBLOCK|O_CLOEXEC) = 5
[   14.158712] testsuite-50.sh[620]: ioctl(5, _IOC(_IOC_NONE, 0x4c, 0xa, 0), 0x7fff0f5d0a70) = -1 EINVAL (Invalid argument)
[   14.158807] testsuite-50.sh[620]: close(5)                                = 0
[   14.158807] testsuite-50.sh[620]: close(4)                                = 0
[   14.158807] testsuite-50.sh[620]: close(3)                                = 0
[   14.158917] testsuite-50.sh[620]: writev(2, [{iov_base="Failed to set up loopback device"..., iov_len=50}, {iov_base="\n", iov_len=1}], 2
[   14.158917] testsuite-50.sh[624]: Failed to set up loopback device: Invalid argument
```

Thanks for digging into this!

Reply:
uh, what? LOOP_CONFIGURE is a kernel 5.8 thing, i am super confused now. it should return ENOTTY or so on 5.6!

Reply:
yup, very confusing, but 100% reproducible:

```
Breakpoint 1, loop_configure (fd=5, c=0x7fffffffe240) at ../src/shared/loop-util.c:39
warning: Source file is more recent than executable.
39	        assert(fd >= 0);
(gdb) n
40	        assert(c);
(gdb) 
42	        if (ioctl(fd, LOOP_CONFIGURE, c) < 0) {
(gdb) 
44	                if (!ERRNO_IS_NOT_SUPPORTED(errno))
(gdb) p errno
$1 = 22
(gdb) q
A debugging session is active.

	Inferior 1 [process 195599] will be killed.

Quit anyway? (y or n) y
$ uname -a
Linux bluca-lenovo 5.6.0-0.bpo.2-amd64 #1 SMP Debian 5.6.14-2~bpo10+1 (2020-06-09) x86_64 GNU/Linux
```

Reply:
did they backport that? strange

Reply:
They did not:

https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/include/uapi/linux/loop.h?h=linux-5.6.y

I guess that's not the first check the kernel does?

Reply:
There we go - fallback to lo_simple_ioctl(): 

https://github.com/torvalds/linux/blob/v5.6/drivers/block/loop.c#L1632

Which then returns EINVAL:

https://github.com/torvalds/linux/blob/v5.6/drivers/block/loop.c#L1588

Reply:
indeed! thanks for tracking this down. bummer!

quite frankly, this looks like a kernel bug to me... they should return ENOTTY instead of EINVAL. Anyone fancies prepping a kernel patch for that?

but either way, we should be able to handle this on our side easily too.

Reply:
Given it would change the ioctl return code for everything it doesn't know about, it sounds like a backward incompatible change to me - so I don't think it would be acceptable, unfortunately :-/